subsys/usb/class/mass_storage.c

```diff
@@ -479,6 +479,11 @@ static bool infoTransfer(void)
 
 static void fail(void)
 {
+	if (cbw.DataLength) {
+		/* Stall data stage */
+		usb_ep_set_stall(mass_ep_data[MSD_IN_EP_IDX].ep_addr);
+	}
```

Comment:
I think sendCSW() should be in the else brunch as it can not be transferred from stalled endpoint anyway: 

```diff
diff --git a/subsys/usb/class/mass_storage.c b/subsys/usb/class/mass_storage.c
index 08cdadf3b8..db6af68eeb 100644
--- a/subsys/usb/class/mass_storage.c
+++ b/subsys/usb/class/mass_storage.c
@@ -479,8 +479,13 @@ static bool infoTransfer(void)
 
 static void fail(void)
 {
-       csw.Status = CSW_FAILED;
-       sendCSW();
+       if (cbw.DataLength) {
+               /* Stall data stage */
+               usb_ep_set_stall(mass_ep_data[MSD_IN_EP_IDX].ep_addr);
+       } else {
+               csw.Status = CSW_FAILED;
+               sendCSW();
+       }
 }

```

Reply:
For this to work I think one would have to add a case for  `USB_DC_CLEAR_HALT` in `mass_storage_status_cb` and send the CSW once the halt is cleared. Otherwise the status won't be sent.

There is already a code path where the endpoint is stalled and the status response written right away:
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/usb/class/mass_storage.c#L574
Which is why I assumed this was a safe approach (I have only tested on an nRF52 device)

Reply:
From USB Spec. 2.0 `The state of a function after returning a STALL (for any endpoint except the default endpoint) is undefined`. Stalling an endpoint and filling the buffer of the same endpoint may not work as expected. A driver/controller may flush the buffer and reset all the properties of the endpoint after ClearFeature(Halt).

Reply:
@jfischer-phytec-iot Maybe something like this would be a cleaner approach, that doesn't depend on the flushing behavior of the driver/controller:

```diff --git a/subsys/usb/class/mass_storage.c b/subsys/usb/class/mass_storage.c
index 08cdadf3b8..c1aecc9dc9 100644
--- a/subsys/usb/class/mass_storage.c
+++ b/subsys/usb/class/mass_storage.c
@@ -480,7 +480,14 @@ static bool infoTransfer(void)
 static void fail(void)
 {
 	csw.Status = CSW_FAILED;
-	sendCSW();
+
+	if (cbw.DataLength) {
+		/* Stall data stage first, then send CSW */
+		stage = MSC_SEND_CSW;
+		usb_ep_set_stall(mass_ep_data[MSD_IN_EP_IDX].ep_addr);
+	} else {
+		sendCSW();
+	}
 }
 
 static void CBWDecode(u8_t *buf, u16_t size)
@@ -838,6 +845,11 @@ static void mass_storage_status_cb(struct usb_cfg_data *cfg,
 		break;
 	case USB_DC_SOF:
 		break;
+	case USB_DC_CLEAR_HALT:
+		if (stage == MSC_SEND_CSW) {
+			sendCSW();
+		}
+		break;
 	case USB_DC_UNKNOWN:
 	default:
 		LOG_DBG("USB unknown state");
```

Reply:
Actually I am not sure about error handling for USB Mass Storage Class. But this code get my attention, I have to review both the code and the specification. But stalling the IN endpoint and report the same error over sendCSW seems not correct to me.

Reply:
I went back to the documentation and here are my thaughts:
> Stalling an endpoint and filling the buffer of the same endpoint may not work as expected

STALL is not a state of an endpoint it's a handshake packet, STALLING an endpoint doesn't mean to change the state of it. It means to send the packet.

> The state of a function after returning a STALL (for any endpoint except the default endpoint) is undefined.

Yes, the state of a function, but the state of the endpoint should stay untouched (it's capable of sending data).

From the usb20 spec:
> **Function**    A USB device that provides a capability to the host, such as an ISDN connection, a digital microphone, or speakers.

Reply:
> STALL is not a state of an endpoint it's a handshake packet, STALLING an endpoint doesn't mean to change the state of it. It means to send the packet.
> 
> > The state of a function after returning a STALL (for any endpoint except the default endpoint) is undefined.

It is also a state of an endpoint, it is stalled until "halt has been cleared through host intervention".

I did not found what endpoint state should be after clear halt, any clues?

Reply:
>It is also a state of an endpoint, it is stalled until "halt has been cleared through host intervention".

That's right, I misunderstood something. Host is clearly doing [this](https://user-images.githubusercontent.com/6489829/75673429-381b3600-5c83-11ea-917b-6a0c61d45b57.png)

> edit: I think correct is "state of endpoint is halted"

It is indeed. This means: "STALL is returned by a function in response to an IN token or after the data phase of an OUT or in response to a PING transaction".
I was refering to
>But stalling the IN endpoint and report the same error over sendCSW seems not correct to me

It says nothing about blocking the CSW. We are stalling the endpoint but we should not block sending the CSW. It has to be send in order to fulfill msc dataflow. From USB msc BBB documentation:

_"The device responds as defined in 6 - Host/Device Data
Transfers by transferring data, STALLing endpoints
when specified, and returning the appropriate CSW."_ - 5.3 of usbmassbulk_10.pdf

Reply:
> It says nothing about blocking the CSW. We are stalling the endpoint but we should not block sending the CSW. It has to be send in order to fulfill msc dataflow. From USB msc BBB documentation:

It is the same IN endpoint, also it is sending CSW over halted endpoint. But what is correct response to unsupported CB? Meaningful CBW and Valid CBW are specified.

Reply:
> 
> 
> > It says nothing about blocking the CSW. We are stalling the endpoint but we should not block sending the CSW. It has to be send in order to fulfill msc dataflow. From USB msc BBB documentation:
> 
> It is the same IN endpoint, also it is sending CSW over halted endpoint. But what is correct response to unsupported CB? Meaningful CBW and Valid CBW are specified.

I think stalling the endpoint is a valid approach when the host expects a data but the command is unsupported by the device:
(from https://www.usb.org/sites/default/files/usbmassbulk_10.pdf) 
```
5.3.2_ Data Transport
All data transport shall begin on a packet boundary.
The host shall attempt to transfer the exact number of bytes to or from the device as specified by thedCBWDataTransferLength and the Direction bit.
The device shall respond as specified in 6 - Host/DeviceData Transfers.

To report an error before data transport completes and to maximize data integrity,
the device may terminate the command by STALLing the endpoint in use (the Bulk-In endpoint during data in, the Bulk-Out endpoint during data out).
```

Reply:
It should STALL indeed. I was wondering if sending csw just after stall is valid approach. From the same spec:

_"If the CBW is not valid, the device shall STALL the Bulk-In pipe. Also, the device shall either STALL the Bulk-Out pipe, or the device shall accept and discard any Bulk-Out data. The device shall maintain this state until a Reset Recovery."_ - 6.6.1 CBW Not Valid

Nothing about sending csw with Phase error (0x02). What is more:

_"The device may not be able to fully satisfy the host's request. At the point when the device discovers that it cannot fully satisfy the request, there may be a Data-In or Data-Out transfer in progress on the bus, and the host may have other pending requests. The device may cause the host to terminate such transfers by STALLing the appropriate pipe. **The response of a device to a CBW that is not meaningful is not specified.**"_ - 6.4 Device Error Handling.

I think @jfischer-phytec-iot is right here and we should follow what You proposed in the commend:
``` diff
index 08cdadf3b8..c1aecc9dc9 100644
--- a/subsys/usb/class/mass_storage.c
+++ b/subsys/usb/class/mass_storage.c
@@ -480,7 +480,14 @@ static bool infoTransfer(void)
 static void fail(void)
 {
 	csw.Status = CSW_FAILED;
-	sendCSW();
+
+	if (cbw.DataLength) {
+		/* Stall data stage first, then send CSW */
+		stage = MSC_SEND_CSW;
+		usb_ep_set_stall(mass_ep_data[MSD_IN_EP_IDX].ep_addr);
+	} else {
+		sendCSW();
+	}
 }
 
 static void CBWDecode(u8_t *buf, u16_t size)
@@ -838,6 +845,11 @@ static void mass_storage_status_cb(struct usb_cfg_data *cfg,
 		break;
 	case USB_DC_SOF:
 		break;
+	case USB_DC_CLEAR_HALT:
+		if (stage == MSC_SEND_CSW) {
+			sendCSW();
+		}
+		break;
 	case USB_DC_UNKNOWN:
 	default:
 		LOG_DBG("USB unknown state");
```
With (maybe) few updates:
``` diff
diff --git a/subsys/usb/class/mass_storage.c b/subsys/usb/class/mass_storage.c
index 08cdadf3b8..81ae14cbab 100644
--- a/subsys/usb/class/mass_storage.c
+++ b/subsys/usb/class/mass_storage.c
@@ -480,7 +480,14 @@ static bool infoTransfer(void)
 static void fail(void)
 {
 	csw.Status = CSW_FAILED;
-	sendCSW();
+
+	if (cbw.DataLength) {
+		/* Stall data stage */
+		stage = MSC_SEND_CSW;
+		usb_ep_set_stall(mass_ep_data[MSD_IN_EP_IDX].ep_addr);
+	} else {
+		sendCSW();
+	}
 }
 
 static void CBWDecode(u8_t *buf, u16_t size)
@@ -808,6 +815,8 @@ static void mass_storage_status_cb(struct usb_cfg_data *cfg,
 	ARG_UNUSED(param);
 	ARG_UNUSED(cfg);
 
+	u8_t ep = (u8_t)*param;
+
 	/* Check the USB status and do needed action if required */
 	switch (status) {
 	case USB_DC_ERROR:
@@ -838,6 +847,12 @@ static void mass_storage_status_cb(struct usb_cfg_data *cfg,
 		break;
 	case USB_DC_SOF:
 		break;
+	case USB_DC_CLEAR_HALT:
+		if (stage == MSC_SEND_CSW &&
+		    ep == mass_ep_data[MSD_IN_EP_IDX].ep_addr) {
+			sendCSW();
+		}
+		break;
 	case USB_DC_UNKNOWN:
 	default:
 		LOG_DBG("USB unknown state");
```

Reply:
> It should STALL indeed. I was wondering if sending csw just after stall is valid approach. From the same spec:
> 
> _"If the CBW is not valid, the device shall STALL the Bulk-In pipe. Also, the device shall either STALL the Bulk-Out pipe, or the device shall accept and discard any Bulk-Out data. The device shall maintain this state until a Reset Recovery."_ - 6.6.1 CBW Not Valid
> 
> Nothing about sending csw with Phase error (0x02). What is more:
> 
> _"The device may not be able to fully satisfy the host's request. At the point when the device discovers that it cannot fully satisfy the request, there may be a Data-In or Data-Out transfer in progress on the bus, and the host may have other pending requests. The device may cause the host to terminate such transfers by STALLing the appropriate pipe. **The response of a device to a CBW that is not meaningful is not specified.**"_ - 6.4 Device Error Handling.
> 
> I think @jfischer-phytec-iot is right here and we should follow what You proposed in the commend

Wait, look at 6.7, there is always something like that:
```
The host shall attempt to receive a CSW.
x. On a STALL condition receiving the CSW, then:
• The host shall clear the Bulk-In pipe.
• The host shall again attempt to receive the CSW.
```

Reply:
I believe ``` clear the Bulk-In pipe. ``` means Reset Recovery here... 

It matches our case, doesn't it?

Reply:
I mean, csw must be send but after Reset Recovery.

Reply:
@nordic-auko I apologize for the long discussion. What I looked for is described in the Section 6.7. I think it is correct to stall and reply with CSW_FAILED. Thanks for the patience.

Reply:
> I mean, csw must be send but after Reset Recovery.

after clear halt. The MSC code do stall and sendCSW on many places, so let us fix it the same way, but we probably keep that in mind.

Reply:
Yes, sorry I wrote this without thinking, after ClearFeature(ENDPOINT_HALT) of course.

Reply:
Thanks for the feedback, @jfischer-phytec-iot  and @emob-nordic 

I'll update the PR to stall the IN endpoint (if there is a data stage) and send CSW after ClearFeature(Halt) event.

Should I also update the other cases where there  is a `usb_ep_set_stall(mass_ep_data[MSD_IN_EP_IDX].ep_addr)` followed by `sendCSW()`?

Reply:
I suggest to keep the patch how it is for now. But than investigate what is the proper approach. I am still looking for the specified endpoint state after clear halt.